Add some technical implementation details for Widgets Customizer (#33026)
WordPress/gutenberg: packages/customize-widgets/README.md

_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for ES2015+ such as IE browsers then using [core-js](https://github.com/zloirock/core-js) will add polyfills for these methods._

## Technical implementation details

> The new Widgets Customizer replaces `Appearance > Customize > Widgets` with block editors. The original Customizer is a Backbone app, but the editor is a React app. One of the challenges is to integrate them together but make sure features from both sides still work.

Reviewer: ```suggestion
The new Widgets Customizer replaces `Appearance > Customize > Widgets` with block-based editors. The original Customizer is a Backbone app, but the new editor is a React app. One of the challenges is to integrate them together but make sure features from both sides still work.
```

Reviewer: Also I think you might like to expand a bit on why we have multiple editors. It could be helpful.

Author: Do you mean why do we have both the Widgets screen and the Widgets Customizer? I'm not really well educated to write that though 😅, and perhaps such thing is better documented in [core article](https://wordpress.org/support/article/wordpress-widgets/) instead. This is not a new thing either, so probably developers are already familiar with this concept.

> We extend Customizer's sections and controls in the `controls` directory and inject some custom logic for the editor. We use React portal to render each editor in its section to reuse most of the styles and scripts provided by the Customizer.

Reviewer: ```suggestion
We extend the Customizer's sections and controls in the `/controls` directory and inject some custom logic for the editor. We use React portal to render each editor in its section to reuse most of the styles and scripts provided by the Customizer.
```

> `components/sidebar-block-editor` is the entry point for each widget area's block editor. `component/sidebar-block-editor/sidebar-adapter.js` is an adapter to talk to the customize API and transform widget objects into widget instances. `components/sidebar-block-editor/use-sidebar-block-editor.js` is a custom React Hook to integrate the adapter into React and handle most of the translations between blocks and widgets. These allow us to implement basic editing features as well as real-time preview in a backward-compatible way.

Reviewer: ```suggestion
`components/sidebar-block-editor` is the entry point for each widget area's block editor. `component/sidebar-block-editor/sidebar-adapter.js` is an adapter to talk to [the Customize API](https://developer.wordpress.org/themes/customize-api/) and transform widget objects into widget instances. 

`components/sidebar-block-editor/use-sidebar-block-editor.js` is a custom React Hook to integrate the adapter into React and handle most of the translations between blocks and widgets. These allow us to implement basic editing features as well as real-time preview in a backwards-compatible way.
```

> Whenever the blocks change, we run through each block to determine if there are created, edited, or deleted blocks. Then, we converted them to their widget counterparts and call the customize API to update them. For React developers, it can be seen as a custom reconciler or a custom renderer for the Customizer. But instead of targeting DOM as the render target, we are targeting WordPress widgets using the customize API.

Reviewer: ```suggestion
Whenever the blocks change, we run through each block to determine if there are created, edited, or deleted blocks. We then convert them to their widget counterparts and call the Customize API to update them. 

For React developers, this can be thought of as a custom reconciler or a custom renderer for the Customizer. But instead of targeting DOM as the render target, we are targeting WordPress widgets using the Customize API.
```

> This is not usually the way the block editor is intended to be used, hence, we have to also implement some missing features such as undo/redo and custom focus control. It is still a goal to make the block editor as easy to integrate into different systems as possible, so the integration in the Widgets Customizer can be a good experience for us to reflect some drawbacks in our current API and potentially improve them in the future.

Reviewer: ```suggestion
This is not the typical way the block editor is intended to be used. As a result, we have to also implement some missing features such as undo/redo and custom focus control. It is still a goal to make the block editor as easy to integrate into different systems as possible, so the integration in the Widgets Customizer can be a good experience for us to reflect some drawbacks in our current API and potentially improve them in the future.
```
